src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,271 @@
+                            .or_insert_with(Default::default);
+
+                        if let Some(new_offset) = position {
+                            if let Some(offset) = value.1 {
+                                if offset < new_offset {
+                                    *value = (val, position);
+                                }
+                            } else {
+                                *value = (val, position);
+                            }
+                        }
+                    }
+                });
+
+                // Produce (key, val) pairs at any complete times.
+                for (cap, map) in values.iter_mut() {
+                    if !input.frontier.less_equal(cap.time()) {
+                        let mut session = output.session(cap);
+                        for (key, val) in map.drain() {
+                            session.give(((key, val), cap.time().clone()))
+                        }
+                    }
+                }
+                // Discard entries, capabilities for complete times.
+                values.retain(|_cap, map| !map.is_empty());
+            }
+        },
+    )
+}
+
+/// Apply a filter followed by a project to an upsert stream.
+fn apply_linear_operators<G>(
+    stream: &Stream<G, (Row, Option<Row>, Timestamp)>,
+    operator: &LinearOperator,
+    src_type: &RelationType,
+) -> (
+    Stream<G, (Row, Option<Row>, Timestamp)>,
+    Stream<G, DataflowError>,
+)
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    // Determine replacement values for unused columns.
+    // This is copied over from applying linear operators to
+    // the non-upsert case.
+    // At this point, the stream consists of (key, key+payload, time)
+    // so it is ok to delete replace unused values in key+payload
+    // but not to replace unused values in key because that would cause
+    // errors in arrange_from_upsert
+    let position_or = (0..src_type.arity())
+        .map(|col| {
+            if operator.projection.contains(&col) {
+                Some(col)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // If a row does not match a predicate whose support lies in the key
+    // columns, it can be tossed away entirely. If a row does not match
+    // a predicate whose support is not contained in the key columns, it
+    // must be replaced with (key, None) in case if there was previously a
+    // row with the same key that matched the predicate.
+    let key_col_len = src_type.keys[0].len();
+    let mut predicates = operator.predicates.clone();
+    let mut key_predicates = Vec::new();
+    predicates.retain(|p| {
+        let key_predicate = p.support().into_iter().all(|c| c < key_col_len);
+        if key_predicate {
+            key_predicates.push(p.clone());
+        }
+        !key_predicate
+    });
+
+    let mut storage = Vec::new();
+    let mut filtered_storage = Vec::new();
```

Comment:
I think this is optional. You can call `ok_session.give()` and the overhead is nominal (the call can be inlined). No particular reason to rip it out either, but I think it just exposes some of the session internals up in to this code (that there is a buffer it accumulates things in to).